udmis/src/main/java/com/google/bos/udmi/service/core/StateHandler.java

```diff
@@ -0,0 +1,87 @@
+import static com.google.udmi.util.GeneralUtils.ifNotNull;
+
+import com.google.bos.udmi.service.messaging.MessageContinuation;
+import com.google.bos.udmi.service.messaging.MessagePipe;
+import com.google.bos.udmi.service.messaging.MessagePipe.HandlerSpecification;
+import com.google.bos.udmi.service.messaging.StateUpdate;
+import com.google.bos.udmi.service.pod.ComponentBase;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.udmi.util.Common;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import udmi.schema.Envelope;
+import udmi.schema.Envelope.SubFolder;
+import udmi.schema.Envelope.SubType;
+import udmi.schema.MessageConfiguration;
+import udmi.schema.State;
+
+/**
+ * Core UDMIS function to process raw State messages from devices and normalize them for
+ * the rest of the system. Involves tagging the envelope with the appropriate designators,
+ * and splitting up the monolithic block into constituent parts.
+ */
+public class StateHandler extends ComponentBase {
+
+  private static final Set<String> STATE_SUB_FOLDERS =
+      Arrays.stream(SubFolder.values()).map(SubFolder::value).collect(Collectors.toSet());
+
+  private final List<HandlerSpecification> messageHandlers = ImmutableList.of(
+      messageHandlerFor(Object.class, this::defaultHandler),
+      messageHandlerFor(Exception.class, this::exceptionHandler),
+      messageHandlerFor(StateUpdate.class, this::messageHandler)
+  );
+
+  final MessagePipe pipe;
+  int exceptionCount;
+  int defaultCount;
+
+  public StateHandler(MessagePipe pipe) {
+    this.pipe = pipe;
+  }
+
+  private void exceptionHandler(Exception e) {
+    exceptionCount++;
+    info("Received processing exception: " + Common.getExceptionMessage(e));
+    e.printStackTrace();
+  }
+
+  private void defaultHandler(Object defaultedMessage) {
+    defaultCount++;
+    MessageContinuation continuation = pipe.getContinuation(defaultedMessage);
+    Envelope envelope = continuation.envelope;
+    Preconditions.checkState(envelope.subType == null, "subType is not null");
+    Preconditions.checkState(envelope.subFolder == null, "subFolder is not null");
+    info("Processing untyped (assumed state) message into state update");
+    envelope.subType = SubType.STATE;
+    envelope.subFolder = SubFolder.UPDATE;
+    continuation.reprocess();
+  }
+
+  private void messageHandler(State message) {
+    info("Sharding state message to pipeline out as incremental updates");
+    Arrays.stream(State.class.getFields()).forEach(field -> {
+      try {
+        if (STATE_SUB_FOLDERS.contains(field.getName())) {
+          ifNotNull(field.get(message), x -> pipe.publish(x));
+        }
+      } catch (Exception e) {
+        throw new RuntimeException("While extracting field " + field.getName(), e);
+      }
+    });
+  }
+
+  public static StateHandler forConfig(MessageConfiguration configuration) {
```

Comment:
Honestly, I never pay attention to the ordering of methods. It's one of those if-the-IDE-does-it-or-the-linter-catches-it then I'll fix it kinds of things. I'd say if there was a linter option you could find that enforced it then I'd be for it, but otherwise I don't see how it practically is enforceable.

Reply:
Ok, I've done my best to put things in a logical order. Honestly it just doesn't seem like something by brain works with (I would stare at the methods and nothing seemed like a good linear order)... I think I generally just don't think of code in a linear way so it kinda trips me up!  PTAL and let me know if you think I should take another pass.